### PR TITLE
OBPIH-6118,6115 Improve preference type and add rating type option endpoints

### PR DESCRIPTION
**Pull request by awalkowiak**

### CODE REVIEW

File: src/js/api/services/ProductSupplierApi.js [hunk 1]
```diff
@@ -5,7 +5,10 @@ import {
 import apiClient from 'utils/apiClient';
 
 export default {
-  getPreferenceTypeOptions: (config) => apiClient.get(PREFERENCE_TYPE_OPTIONS, config),
+  getPreferenceTypeOptions: (config) => apiClient.get(
+    PREFERENCE_TYPE_OPTIONS('?includeMultiple=true&includeNone=true'),
+    config,
+  ),
```
Comment (alannadolny) [inline src/js/api/services/ProductSupplierApi.js:11]: I think you should pass those params within the `config` object
```
params: {
   includeMultiple: true,
   includeNone: true,
}
```

File: grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy [hunk 2]
```diff
@@ -80,24 +81,40 @@ class SelectOptionsApiController {
...
+        if (includeMultiple) {
+            preferenceTypeOptions.add([
                 id: ProductSupplierService.PREFERENCE_TYPE_MULTIPLE,
                 label: g.message(code: "react.productSupplier.preferenceType.multiple.label", default: "Multiple")
-            ],
-            [
+            ])
+        }
+
+        boolean includeNone = params.boolean("includeNone") ?: false
```
Comment (drodzewicz) [inline grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy:94]:
```suggestion
        boolean includeNone = params.boolean("includeNone", false)
```

File: grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy [hunk 3]
```diff
@@ -80,24 +81,40 @@ class SelectOptionsApiController {
     }
 
     def preferenceTypeOptions() {
-        List<Map<String, String>> defaultPreferenceTypes = [
-            [
+        List<Map<String, String>> preferenceTypeOptions = []
+
+        boolean includeMultiple = params.boolean("includeMultiple") ?: false
```
Comment (drodzewicz) [inline grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy:86]:
```suggestion
        boolean includeMultiple = params.boolean("includeMultiple", false)
```

**Review by drodzewicz — CHANGES_REQUESTED** [review]

File: src/js/api/urls.js [hunk 4]
```diff
@@ -67,7 +67,7 @@ export const PAYMENT_TERMS_OPTION = `${API}/paymentTermOptions`;
 export const USERS_OPTIONS = `${API}/users`;
 
 // PREFERENCE TYPES
-export const PREFERENCE_TYPE_OPTIONS = `${API}/preferenceTypeOptions`;
+export const PREFERENCE_TYPE_OPTIONS = (params) => `${API}/preferenceTypeOptions${params}`;
```
Comment (alannadolny) [inline src/js/api/urls.js:70]: After changing the `config` object, you shouldn't pass the params argument

File: grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy [hunk 5]
```diff
@@ -80,24 +81,40 @@ class SelectOptionsApiController {
...
-        defaultPreferenceTypes.addAll(preferenceTypes)
 
-        render([data: defaultPreferenceTypes] as JSON)
+        preferenceTypeOptions.addAll(preferenceTypes)
+
+        render([data: preferenceTypeOptions] as JSON)
+    }
+
+    def ratingTypeCodeOptions() {
+        List ratingTypeCodeOptions = RatingTypeCode.list().collect {
```
Comment (alannadolny) [inline grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy:115]: Make this `List` type parametrized

**Review by alannadolny — CHANGES_REQUESTED** [review]

File: src/js/api/services/ProductSupplierApi.js [hunk 6]
```diff
@@ -5,7 +5,10 @@ import {
 import apiClient from 'utils/apiClient';
 
 export default {
-  getPreferenceTypeOptions: (config) => apiClient.get(PREFERENCE_TYPE_OPTIONS, config),
+  getPreferenceTypeOptions: (config) => apiClient.get(
+    PREFERENCE_TYPE_OPTIONS('?includeMultiple=true&includeNone=true'),
```
Comment (kchelstowski) [inline src/js/api/services/ProductSupplierApi.js:9]: the params should be in the config. You would need to change this:
```js
const config = {
    signal: controller.signal,
};
```
in the `useProductSupplierFilters.jsx`
to include params:
```js
const config = {
    signal: controller.signal,
    params: {
      includeMultiple: true,
      includeNone: true,
    },
};
```

File: grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy [hunk 7]
```diff
@@ -80,24 +81,40 @@ class SelectOptionsApiController {
...
+            [id: it.name, label: g.message(code: "enum.RatingTypeCode.$it.name", default: "$it.name")]
```
Comment (kchelstowski) [inline grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy:116]: nitpicky: I think the default could be simplified to: `default: it.name`

**Review by kchelstowski — CHANGES_REQUESTED** [review]

File: grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy (hunk 7 above)
Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy:116]: yup nitpicky

Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/api/SelectOptionsApiController.groovy]: fixed

File: src/js/api/services/ProductSupplierApi.js (hunk 6 above)
Comment (awalkowiak) [reply src/js/api/services/ProductSupplierApi.js]: fixed

File: src/js/api/services/ProductSupplierApi.js (hunk 1 above)
Comment (awalkowiak) [reply src/js/api/services/ProductSupplierApi.js]: fixed

File: src/js/api/urls.js (hunk 4 above)
Comment (awalkowiak) [reply src/js/api/urls.js]: fixed

File: src/js/actions/index.js [hunk 8]
```diff
@@ -679,3 +681,11 @@ export const fetchPreferenceTypes = (config) => async (dispatch) => {
     payload: preferenceTypes?.data?.data,
   });
 };
+
+export const fetchRatingTypeCodes = () => async (dispatch) => {
+  const ratingTypeCodes = await apiClient.get(RATING_TYPE_OPTIONS);
```
Comment (alannadolny) [inline src/js/actions/index.js:686]: I feel like we should use our API services
(I mean those services :point_down:)
image

**Review by alannadolny — CHANGES_REQUESTED** [review]

**Review by kchelstowski — CHANGES_REQUESTED** [review]
one small thing

File: src/js/hooks/productSupplier/form/useProductSupplierForm.js [hunk 9]
```diff
@@ -11,6 +15,17 @@ const useProductSupplierForm = () => {
...
+  const dispatch = useDispatch();
+
+  const {
+    ratingTypeCodes,
+  } = useSelector((state) => ({
+    ratingTypeCodes: state.productSupplier.ratingTypeCodes,
+  }));
+
+  useEffect(() => {
+    dispatch(fetchRatingTypeCodes());
```
[the remark this answers is not in the document]
Comment (awalkowiak) [reply src/js/hooks/productSupplier/form/useProductSupplierForm.js:28]: done

File: src/js/actions/index.js (hunk 8 above)
Comment (awalkowiak) [reply src/js/actions/index.js]: done

**Review by kchelstowski — APPROVED** [review]

**Review by drodzewicz — APPROVED** [review]

**Pull request merged**